Fix bounds check in Queue::writeBuffer when buffer has been destroyed

achristensen07 opened the pull request:
#### 7fed8609a81438d9ae188d1b83db749dba59236d
<pre>
Fix bounds check in Queue::writeBuffer when buffer has been destroyed
<a href="https://bugs.webkit.org/show_bug.cgi?id=270538">https://bugs.webkit.org/show_bug.cgi?id=270538</a>
<a href="https://rdar.apple.com/123811082">rdar://123811082</a>

Reviewed by Mike Wyrzykowski.

Buffer::destroy updates m_buffer but didn&apos;t update m_size, and it shouldn&apos;t update the size as
seen by the JS API.

Instead of updating it, I derive the size from m_buffer and replace size() with currentSize()
and initialSize()

I also add a null check to avoid writing to a buffer that is null.

* LayoutTests/fast/webgpu/write-to-destroyed-buffer-expected.txt: Added.
* LayoutTests/fast/webgpu/write-to-destroyed-buffer.html: Added.
* Source/WebGPU/WebGPU/Buffer.h:
(WebGPU::Buffer::create):
* Source/WebGPU/WebGPU/Buffer.mm:
(WebGPU::Device::createBuffer):
(WebGPU::Buffer::Buffer):
(WebGPU::Buffer::getMappedRange):
(WebGPU::Buffer::errorValidatingMapAsync const):
(WebGPU::Buffer::mapAsync):
(WebGPU::Buffer::size const):
* Source/WebGPU/WebGPU/Queue.mm:
(WebGPU::Queue::validateWriteBuffer const):

Canonical link: <a href="https://commits.webkit.org/275756@main">https://commits.webkit.org/275756@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e6165e8a38d320dcaa9532c372fdfe37784090cf

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/42717 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/21738 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/45118 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/45329 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/38840 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/25403 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/19103 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/35359 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/43290 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/18748 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/36772 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/16315 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/16380 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/37833 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/773 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/38893 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/38164 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/46834 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/17535 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/14458 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/42076 "Passed tests") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/19154 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/37102 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/40703 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/19333 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5782 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/18799 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski reviewed (changes requested):
I think this regresses https://gpuweb.github.io/cts/standalone/?q=webgpu:api,validation,buffer,mapping:getMappedRange,sizeAndOffsetOOB,mapped:* and a few other tests

mwyrzykowski reviewed (changes requested):
It seems https://gpuweb.github.io/cts/standalone/?q=webgpu:api,validation,buffer,mapping:getMappedRange,sizeAndOffsetOOB,mapped:* still fails after this change but passes on ToT WebKit (main-275701@main-1709666740_dbef77526e5f)

mwyrzykowski commented on Source/WebGPU/WebGPU/Queue.mm:
```diff
@@ -278,14 +278,17 @@
     if (!(buffer.usage() & WGPUBufferUsage_CopyDst))
         return false;
 
+    if (!buffer.buffer().contents)
```
I think this check might be problematic.

mwyrzykowski replied:
I think a private buffer which we create in some cases will have a nullptr contents so this check is only valid if `buffer.buffer().storageMode != MTLStorageModePrivate`

mwyrzykowski commented on Source/WebGPU/WebGPU/Queue.mm:
```diff
@@ -278,14 +278,17 @@
     if (!(buffer.usage() & WGPUBufferUsage_CopyDst))
         return false;
 
+    if (buffer.buffer().storageMode != MTLStorageModePrivate && !buffer.buffer().contents)
```
I think we also need a !buffer.isDestroyed() check here like so:
```
if (!buffer.isDestroyed() && buffer.buffer().storageMode != MTLStorageModePrivate && !buffer.buffer().contents)
```
It appears this test fails after this change:

https://gpuweb.github.io/cts/standalone/?q=webgpu:api,validation,encoding,cmds,clearBuffer:buffer_state:bufferState=%22destroyed%22

mwyrzykowski reviewed (changes requested):
Sorry I found another failing test: https://gpuweb.github.io/cts/standalone/?q=webgpu:api,validation,encoding,cmds,clearBuffer:buffer_state:bufferState=%22destroyed%22

We really want to enable all of the passing ones so EWS can do this for us, but we need to go through and enable the ones which don't take minutes to run

mwyrzykowski reviewed (commented):
Do you think we should limit this change to skipping the memcpy if buffer.buffer().contents is nullptr? That is unlikely to regress the existing tests while fixing the crash

mwyrzykowski reviewed (approved):
Looks great - thank you!!

achristensen07 committed:
Fix bounds check in Queue::writeBuffer when buffer has been destroyed
https://bugs.webkit.org/show_bug.cgi?id=270538
rdar://123811082

Reviewed by Mike Wyrzykowski.

Buffer::destroy updates m_buffer but didn't update m_size, and it shouldn't update the size as
seen by the JS API.

Instead of updating it, I derive the size from m_buffer and replace size() with currentSize()
and initialSize()

I also add a null check to avoid writing to a buffer that is null.

* LayoutTests/fast/webgpu/write-to-destroyed-buffer-expected.txt: Added.
* LayoutTests/fast/webgpu/write-to-destroyed-buffer.html: Added.
* Source/WebGPU/WebGPU/Buffer.h:
(WebGPU::Buffer::create):
* Source/WebGPU/WebGPU/Buffer.mm:
(WebGPU::Device::createBuffer):
(WebGPU::Buffer::Buffer):
(WebGPU::Buffer::getMappedRange):
(WebGPU::Buffer::errorValidatingMapAsync const):
(WebGPU::Buffer::mapAsync):
(WebGPU::Buffer::size const):
* Source/WebGPU/WebGPU/Queue.mm:
(WebGPU::Queue::validateWriteBuffer const):

Canonical link: https://commits.webkit.org/275756@main
